Projet/src/tar.c

```diff
@@ -153,4 +153,129 @@ bool tar_create_folder(struct tar_header* header)
     }
 
     return mkdir(header->name, strtol(header->mode, NULL, 8));
+}
+
+void tar_generate_archive(FILE* archive, char* filenames[], int nb_files)
+{
+    for(int i = 0; i < nb_files; i++) 
+    {
+        FILE* file;
+        
+        if(!(file = fopen(filenames[i], "r")) )
+        {
+            fprintf(stderr, "Error : tar_generate_archive() : cannot open file %s\n", filenames[i]);
+            return;
+        }
+
+        tar_add_file_to_archive(archive, file, filenames[i]);
+
+        fclose(file);
+    }
+
+    tar_add_end_of_file(archive);
+}
+
+bool tar_add_file_to_archive(FILE* archive, FILE* file, char* filename) 
+{
+    struct tar_header header = tar_fill_header(file, filename);
+
+    if(tar_header_is_empty(&header)) 
+    {
+        return false;
+    }
+
+    int* temp = malloc(BLOCKSIZE);
+
+    long int file_size = strtol(header.size, NULL, 8);
+
+    fwrite(&header, BLOCKSIZE, 1, archive);
+
+    for (long int i = 0; i < file_size; i += BLOCKSIZE)
+    {
+        fread(temp, BLOCKSIZE, 1, file);
+        fwrite(temp, BLOCKSIZE, 1, archive);            
+    }
+    
+    free(temp);
+
+    return true;
+}
+
+struct tar_header tar_fill_header(FILE* file, char* filename)
+{
+    struct tar_header header;
+
+    // mode
+    int fd = fileno(file);
+    struct stat file_stat;
+
+    if(fstat(fd, &file_stat) != 0) 
+    {
+        printf("Cest nul\n");
+        return header;
+    }
+
+    // name
+    strncpy(header.name, filename, 100);
+    header.name[sizeof(header.name) - 1] = '\0';
+
+    sprintf(header.mode, "%o", file_stat.st_mode);
+    sprintf(header.owner, "%o", file_stat.st_uid);
+    sprintf(header.group, "%o", file_stat.st_gid);
+    sprintf(header.size, "%lo", file_stat.st_size);
+    sprintf(header.mtime, "%lo", file_stat.st_mtime);
+
+    memset(header.checksum, ' ', sizeof(header.checksum));
+
+    header.type = '0'; // TODO : changer le type
+
+    unsigned int checksum = tar_calculate_checksum_header((char *) &header);
+
+    sprintf(header.checksum, "%06o\0 ", checksum);
+
+    return header;
+}
+
+void tar_add_end_of_file(FILE* archive) 
+{
+    char data[BLOCKSIZE];
+
+    memset(data, 0, BLOCKSIZE);
+
+    for(int i = 0; i < 3; i++)
+    {
+        fwrite(data, BLOCKSIZE, 1, archive);
+    }
+}
+
+// Put the pointer to the beginning of the end blocks of archive
+void tar_seek_to_archive_end(FILE* file) 
+{
+    fseek(file, BLOCKSIZE, SEEK_END);
+
+    struct tar_header header;
+
+    do {
+        tar_read_header(file, &header);
+
+        fseek(file, -BLOCKSIZE * 2, SEEK_CUR);
```

Comment:
2 -> magic number

Reply:
For this one, it's a question of business logic: we seek from 1 to the right when reading, and we seek from 2 to the left (1 because we read a header, and 1 to go to the block from before)
This magic number is therefore normal + allows a better understanding of the code

(I'm still going to add a little comment)